ServerCore/Pages/Puzzles/Edit.cshtml.cs

```diff
@@ -120,6 +120,41 @@ public async Task<IActionResult> OnPostAsync(int puzzleId)
             {
                 await _context.SaveChangesAsync();
 
+                // Check to see if IsForSinglePlayer has changed.
+                // Delete appropriate rows if so
+                if (oldPuzzleView.IsForSinglePlayer != Puzzle.IsForSinglePlayer)
+                {
+                    if (oldPuzzleView.IsForSinglePlayer)
+                    {
+                        // If moving from SinglePlayerPuzzle to team puzzle, delete single player puzzle states
+                        IEnumerable<SinglePlayerPuzzleHintStatePerPlayer> hintStatesToRemove = _context.SinglePlayerPuzzleHintStatePerPlayer.Where(hintState => hintState.Hint.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleHintStatePerPlayer.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleStatePerPlayer> puzzleStatesToRemove = _context.SinglePlayerPuzzleStatePerPlayer.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleStatePerPlayer.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleSubmission> submissionsToRemove = _context.SinglePlayerPuzzleSubmissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleSubmissions.RemoveRange(submissionsToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleUnlockState> unlockStatesToRemove = _context.SinglePlayerPuzzleUnlockStates.Where(unlockState => unlockState.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleUnlockStates.RemoveRange(unlockStatesToRemove);
+                    }
+                    else
+                    {
+                        // If moving from team puzzle to SinglePlayerPuzzle, delete team puzzle states
+                        IEnumerable<HintStatePerTeam> hintStatesToRemove = _context.HintStatePerTeam.Where(hintState => hintState.Hint.PuzzleID == oldPuzzleView.ID);
+                        _context.HintStatePerTeam.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<PuzzleStatePerTeam> puzzleStatesToRemove = _context.PuzzleStatePerTeam.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.PuzzleStatePerTeam.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<Submission> submissionsToRemove = _context.Submissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.Submissions.RemoveRange(submissionsToRemove);
+                    }
+
+                    await _context.SaveChangesAsync();
```

Comment:
Is there any reason this new block shouldn't just come before the existing SaveChangesAsync, removing the need for a second database save?

Reply:
Unless there's a reason why the new code should come first (and please document why if so), I'd just use the same if/then for both blocks of code (old line 163, new line 198) for a) maximum clarity and b) full atomicity of the DB change.

Reply:
Done!